packages/astro/src/core/render-context.ts

```diff
@@ -80,22 +91,51 @@ export class RenderContext {
 	 * - fallback
 	 */
 	async render(componentInstance: ComponentInstance | undefined): Promise<Response> {
-		const { cookies, middleware, pathname, pipeline, routeData } = this;
+		const { cookies, middleware, pathname, pipeline } = this;
 		const { logger, routeCache, serverLike, streaming } = pipeline;
 		const props = await getProps({
 			mod: componentInstance,
-			routeData,
+			routeData: this.routeData,
 			routeCache,
 			pathname,
 			logger,
 			serverLike,
 		});
 		const apiContext = this.createAPIContext(props);
 
-		const lastNext = async () => {
-			switch (routeData.type) {
+		this.counter++;
+		if (this.counter == 4) {
+			return new Response('Loop Detected', {
+				// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/508
+				status: 508,
+				statusText: 'Loop Detected',
+			});
+		}
```

Comment:
```suggestion
		this.counter++;
		if (this.counter === 4) {
			return new Response('Loop Detected', {
				// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/508
				status: 508,
				statusText: 'Loop Detected',
			});
		}
```

Should we also elaborate what kind of loop is happening here?

Reply:
Yeah, for now, the loop detection is very rudimental, and I would be happy to make it a bit smarter in the next release.

Where do you think I should add a better message? In the `statusText`?

Reply:
Yeah having `statusText` a longer description of the issue would be nice. Is this only possibly triggered because of rerouting? Perhaps we can mention that it's caused by that feature so it's easier to debug.